src/core/pointcloud/qgsremotecopcpointcloudindex.cpp

```diff
@@ -0,0 +1,373 @@
+    return nullptr;
+
+  QEventLoop loop;
+  connect( blockRequest.get(), &QgsPointCloudBlockRequest::finished, &loop, &QEventLoop::quit );
+  loop.exec();
+
+  if ( !blockRequest->block() )
+  {
+    QgsDebugMsg( QStringLiteral( "Error downloading node %1 data, error : %2 " ).arg( n.toString(), blockRequest->errorStr() ) );
+  }
+
+  return blockRequest->block();
+}
+
+QgsPointCloudBlockRequest *QgsRemoteCopcPointCloudIndex::asyncNodeData( const IndexedPointCloudNode &n, const QgsPointCloudRequest &request )
+{
+  if ( !fetchNodeHierarchy( n ) )
+    return nullptr;
+  QMutexLocker locker( &mHierarchyMutex );
+
+  // we need to create a copy of the expression to pass to the decoder
+  // as the same QgsPointCloudExpression object might be concurrently
+  // used on another thread, for example in a 3d view
+  QgsPointCloudExpression filterExpression = mFilterExpression;
+  QgsPointCloudAttributeCollection requestAttributes = request.attributes();
+  requestAttributes.extend( attributes(), filterExpression.referencedAttributes() );
+  auto [ blockOffset, blockSize ] = mHierarchyNodePos[n];
+  int pointCount = mHierarchy[n];
+
+  QgsCopcPointCloudBlockRequest *req = new QgsCopcPointCloudBlockRequest( n, mUrl.toString(), attributes(), requestAttributes,
+      scale(), offset(), filterExpression,
+      blockOffset, blockSize, pointCount, mCopcHeaderData, mExtraBytesData );
+  req->startRequest();
+  return req;
+}
+
+bool QgsRemoteCopcPointCloudIndex::hasNode( const IndexedPointCloudNode &n ) const
+{
+  return fetchNodeHierarchy( n );
+}
+
+bool QgsRemoteCopcPointCloudIndex::fetchNodeHierarchy( const IndexedPointCloudNode &n ) const
+{
+  QMutexLocker locker( &mHierarchyMutex );
+
+  QVector<IndexedPointCloudNode> ancestors;
+  IndexedPointCloudNode foundRoot = n;
+  while ( !mHierarchy.contains( foundRoot ) )
+  {
+    ancestors.push_front( foundRoot );
+    foundRoot = foundRoot.parentNode();
+  }
+  ancestors.push_front( foundRoot );
+  for ( IndexedPointCloudNode n : ancestors )
+  {
+    if ( !mHierarchy.contains( n ) )
+      return false;
+    int nodesCount = mHierarchy[n];
+    if ( nodesCount < 0 )
+    {
+      fetchHierarchyPage( mHierarchyNodePos[n].first, mHierarchyNodePos[n].second );
+    }
+  }
+  return mHierarchy.contains( n );
+}
+
+bool QgsRemoteCopcPointCloudIndex::isValid() const
+{
+  return mIsValid;
+}
+
+void QgsRemoteCopcPointCloudIndex::fetchHierarchyPage( uint64_t offset, uint64_t byteSize ) const
+{
+  QNetworkRequest nr( mUrl );
+  nr.setAttribute( QNetworkRequest::HttpPipeliningAllowedAttribute, true );
+  QByteArray queryRange = QStringLiteral( "bytes=%1-%2" ).arg( offset ).arg( offset + byteSize - 1 ).toLocal8Bit();
+  nr.setRawHeader( "Range", queryRange );
+  QgsBlockingNetworkRequest req;
+  QgsBlockingNetworkRequest::ErrorCode errCode = req.get( nr );
+  if ( errCode != QgsBlockingNetworkRequest::NoError )
+  {
+    QgsDebugMsg( QStringLiteral( "Request failed: " ) + mUrl.toString() );
```

Comment:
we should probably return here and not rely on the fact that the reply has empty content - it could be that server sends some response anyway, and we end up writing some wrong data to the hierarchy...